libs/core/langchain_core/chat_history.py

```diff
@@ -64,18 +78,38 @@ def add_ai_message(self, message: Union[AIMessage, str]) -> None:
         else:
             self.add_message(AIMessage(content=message))
 
-    @abstractmethod
     def add_message(self, message: BaseMessage) -> None:
         """Add a Message object to the store.
 
         Args:
             message: A BaseMessage object to store.
         """
-        raise NotImplementedError()
+        if type(self).add_messages != BaseChatMessageHistory.add_messages:
+            # This means that the sub-class has implemented an efficient add_messages
+            # method, so we should usage of add_message to that.
+            self.add_messages([message])
```

Comment:
This creates an infinite loop if the child class only implements `add_messages` and not `add_message`

Reply:
If i'm not mistaken, the conditional immediately above prevents an infinite loop, it's based on checking whether add_messages has been overloaded or not. 

Check the included unit tests that verifies that things work properly when either add_messages or add_message are implemented

Reply:
Oh yeah, you're right! Sorry, I mistook the class, so it won't be a breaking change. The only two cases it currently wouldn't catch are if someone implemented `add_messages` to call `add_message` without overriding `add_message` (e.g., to simply add logging) or used the class as-is (overriding `clear`, of course), calling only `add_messages` and `add_message`. Either of those would result in infinite loops rather than raising `NotImplementedError`. It makes sense to me to either ensure that `NotImplementedError` would always be raised or `@abstractmethod` would be added back to `add_message` so it is abundantly clear what methods are abstract methods.

Reply:
Yeah :( but that's a breaking change for all implementations. if I could make a breaking change, i'd delete `add_message` entirely :). 

Messages should not be added one at a time since production use cases, writing a message usually involves a network request, and the round trips on network requests become expensive quickly.